### PR TITLE
Insights gateway WIP

### PR DESCRIPTION
**Pull request by iNecas**

### CODE REVIEW

**Comment by iNecas** [comment]
Hey, @jhjaggars, @chambridge, @martinkunc, @mfojtik, I'm planning to send an enhancement proposal for introducing the insights-gateway to proxy requests to cloud.redhat.com. Mind giving it a pre-flight check before going for full review in oc/enhancements repo?

I'm mainly looking for factual issues in the proposal, as well as ideas for additional clarification or anything that would decrese the chances of the proposal to be accepted eventually.

We can leave the detailed review for the main repository: this is mainly for giving you some additional time to pivot before going mainstream :)

File: enhancements/insights/insights_gateway.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+## Motivation
+
+In the last couple of releases, we've seen multiple new services build in 
+cloud.redhat.com (c.r.c.) providing additional insights from various areas,
+each needing additional data to be sent from the OCP clusters, such as:
+
+- [cost management](https://github.com/project-koku/korekuta-operator)
+- [subscription watch](https://github.com/chambridge/subscription-watch-operator)
+- [marketplace](https://github.com/RedHatInsights/marketplace-processor)
```
Comment (chambridge) [inline enhancements/insights/insights_gateway.md]: This is the marketplace operator, what you linked is the service on cloud.redhat.com that processes the marketplace content:
https://github.com/redhat-marketplace/redhat-marketplace-operator/blob/acf4f3aaef03f15b6a88b0e2dcf8c3b52c3d7294/pkg/reporter/uploader.go

**Review by chambridge — COMMENTED** [review]
Link to marketplace operator provided.

File: enhancements/insights/insights_gateway.md (hunk 1 above)
Comment (iNecas) [reply enhancements/insights/insights_gateway.md]: Updated, thanks.

File: enhancements/insights/insights_gateway.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  kind: ClusterRole
+  name: rhinsights-upload
+subjects:
+  - kind: ServiceAccount
+    name: olm-operator
+    namespace: olm-operator
+```
+4. before forwarding the request, the io will:
+  - log the event of forwarding the request (service account info, content_type, payload size…)
+  - add additional headers to the request about it's origin (service account etc… - TBD)
```
Comment (iNecas) [inline enhancements/insights/insights_gateway.md]: @jhjaggars what kind of headers you think would be useful here for the ingress service?

File: enhancements/insights/insights_gateway.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+## Drawbacks
+
+1. This proposal partly extends the responsibilities of insights-operator beyond
+   the realm of support.
+2. Problematic behaviour of an external operator (large payloads, too often) can negatively
+   influence the whole insights operator.
+3. There are other components in the OCP cluster (e.g. telemeter) that are sending
+   data to managed services. This proposal is not covering these cases.
+
```
Comment (martinkunc) [inline enhancements/insights/insights_gateway.md:184]: One drawback which comes to my mind is, when an operator was using custom proxy (Like Insights or Monitoring do). In cases when customer would like to have a cluster behind proxy and explicitly not define cluster proxy, but rather enable it only for a particular operator. Like it was the case why it was implemented in Insighst Operator.
Using RBAC roles, if I understand it well, the customer would have to make sure that no new binding refers the gateway service account and eventually remove it.

Comment (iNecas) [reply enhancements/insights/insights_gateway.md:184]: I wonder if this really is a drawback, especially as all the cases we're targeting are similar in a way of sending data to c.r.c.: I can't think of a real case where one would want to have a proxy for one kind of playload and not one for the other.

Regarding opting out specific operator: yes - the RBAC model would be the way to control what component can use the proxy.

**Comment by martinkunc** [comment]
I think it looks good. Maybe it could indicate what potential benefits will operators gain by replacing direct upload by the gateway, because actually it is not too big effort to authenticate with the Token.
I can imagine for example better configuration driven proxy (upload) settings,traffic shaping, or some upload queue. Is there something else ?

**Comment by iNecas** [comment]
@martinkunc incorporated, opened https://github.com/openshift/enhancements/pull/447

**Pull request closed without merging**